akka-typed/src/main/scala/akka/typed/ScalaProcess.scala

```diff
@@ -0,0 +1,538 @@
+     * input for the next computation.
+     */
+    def flatMap[T](f: Out ⇒ Operation[S, T]): Operation[S, T] = FlatMap(this, f)
+
+    /**
+     * Map the value computed by this process step by the given function,
+     * flattening the result if it is an [[Operation]] (by executing the
+     * operation and using its result as the mapped value).
+     *
+     * The reason behind flattening when possible is to allow the formulation
+     * of infinite process loops (as performed for example by server processes
+     * that respond to any number of requests) using for-comprehensions.
+     * Without this flattening a final pointless `map` step would be added
+     * for each iteration, eventually leading to an OutOfMemoryError.
+     */
+    def map[T, Mapped](f: Out ⇒ T)(implicit ev: MapAdapter[S, T, Mapped]): Operation[S, Mapped] = flatMap(ev.lift(f))
+
+    /**
+     * Unfortunately for-comprehensions desugar to `.foreach` if the `yield`
+     * keyword is absent, but they do so throughout the whole expression.
+     * Coding a normal signature for `foreach` would break everything after
+     * the first step as it would not be executed—the `Unit` result is there
+     * already.
+     */
+    def foreach[T, Mapped](f: Out ⇒ T)(implicit ev: MapAdapter[S, T, Mapped]): Operation[S, Mapped] = flatMap(ev.lift(f))
+
+    /**
+     * Only continue this process if the given predicate is fulfilled, terminate
+     * it otherwise.
+     */
+    def filter(p: Out ⇒ Boolean): Operation[S, Out] = flatMap(o ⇒ if (p(o)) Return(o) else ShortCircuit)
+
+    /**
+     * Only continue this process if the given predicate is fulfilled, terminate
+     * it otherwise.
+     */
+    def withFilter(p: Out ⇒ Boolean): Operation[S, Out] = flatMap(o ⇒ if (p(o)) Return(o) else ShortCircuit)
+
+    /**
+     * Wrap as a [[Process]] with infinite timeout and a mailbox capacity of 1.
+     * Small processes that are called or chained often interact in a fully
+     * sequential fashion, where these defaults make sense.
+     */
+    def named(name: String): Process[S, Out] = Process(name, Duration.Inf, 1, this)
+
+    /**
+     * Wrap as a [[Process]] with the given mailbox capacity and infinite timeout.
+     */
+    def withMailboxCapacity(mailboxCapacity: Int): Process[S, Out] = named("").withMailboxCapacity(mailboxCapacity)
+
+    /**
+     * Wrap as a [[Process]] with the given timeout and a mailbox capacity of 1.
+     */
+    def withTimeout(timeout: Duration): Process[S, Out] = named("").withTimeout(timeout)
+
+    /**
+     * Wrap as a [[Process]] but without a name and convert to a [[Behavior]].
+     */
+    def toBehavior: Behavior[ActorCmd[S]] = named("").toBehavior
+
+  }
+
+  /*
+   * These are the private values that make up the core algebra.
+   */
+
+  private[typed] case class FlatMap[S, Out1, Out2](first: Operation[S, Out1], then: Out1 ⇒ Operation[S, Out2]) extends Operation[S, Out2] {
+    override def toString: String = s"FlatMap($first)"
```

Comment:
not sure I like hiding the 2nd param. Often confuses people "wat, I thought it had 2 params!".
Perhaps `s"FlatMap($first, <function1:then>)`?